core/src/main/java/org/pentaho/platform/dataaccess/datasource/api/DatasourceService.java

```diff
@@ -154,4 +156,48 @@ public static boolean isDSWDatasource( Domain domain ) {
     }
     return !isMetadataDatasource( domain );
   }
+
+  protected String isMetaDataSource( String id ) {
+    Domain domain;
+    try {
+      domain = metadataDomainRepository.getDomain( id );
+      if ( domain == null ) {
+        return null;
+      }
+    } catch ( Exception e ) { // If we can't load the domain then we MUST return false
+      return null;
+    }
+    return isMetadataDatasource( domain ) ? id : null;
+  }
+
+  protected int noOfThreads() {
+    IPluginResourceLoader resLoader = PentahoSystem.get( IPluginResourceLoader.class, null );
+    String noOfThreadsAsString = null;
+    int noOfThreads = 1;
+    try {
+      noOfThreadsAsString =
+              resLoader.getPluginSetting( getClass(), "settings/threads-for-user-session-cache" ); //$NON-NLS-1$
+    } catch ( Exception e ) {
+      e.printStackTrace();
+    }
+    if ( StringUtils.isNotBlank( noOfThreadsAsString ) ) {
+      noOfThreads = Integer.parseInt( noOfThreadsAsString );
```

Comment:
We shouldn't assume that valid input is entered.
We need to handle the cases that:
< 1 entered
noOfThreads > `Runtime.getRuntime().availableProcessors();`
Integer.parseInt(string) throws NumberFormatException
https://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html#parseInt(java.lang.String)